includes/class-wc-payments-payment-request-button-handler.php

```diff
@@ -1703,4 +1707,46 @@ private function get_taxes( $product, $price ) {
 		// Normally there should be a single tax, but `calc_tax` returns an array, let's use it.
 		return WC_Tax::calc_tax( $price, $rates, false );
 	}
+
+	/**
+	 * Returns true if the given product is a subscription that has free trial and requires shipping.
+	 * This could be a subscription product with a trial period or a synchronised subscription with a delayed payment.
+	 *
+	 * If the product is a variable subscription, this function will return true if all of its variations have a trial and require shipping.
+	 *
+	 * @since 6.8.0
+	 *
+	 * @param WC_Product|null $product Product object.
+	 *
+	 * @return boolean
+	 */
+	public function product_has_trial_and_needs_shipping( $product ) {
+		if ( ! class_exists( 'WC_Subscriptions_Product' ) || ! class_exists( 'WC_Subscriptions_Synchroniser' ) || ! WC_Subscriptions_Product::is_subscription( $product ) ) {
+			return false;
+		}
+
+		if ( $product->get_type() === 'variable-subscription' ) {
+			$products = $product->get_available_variations( 'object' );
+		} else {
+			$products = [ $product ];
+		}
+
+		foreach ( $products as $product ) {
+			// Return early if the product doesn't require shipping.
+			if ( ! $product->needs_shipping() ) {
+				return false;
+			}
+
+			// If product is synced, check if the first payment is upfront or today (i.e. no trial period). If product is not synced, check if it has a trial period.
+			if ( WC_Subscriptions_Synchroniser::is_product_synced( $product ) ) {
+				if ( WC_Subscriptions_Synchroniser::is_payment_upfront( $product ) || WC_Subscriptions_Synchroniser::is_today( WC_Subscriptions_Synchroniser::calculate_first_payment_date( $product, 'timestamp' ) ) ) {
```

Comment:
```suggestion
				if ( WC_Subscriptions_Synchroniser::is_payment_upfront( $product ) ) {
```

You can get rid of the `WC_Subscriptions_Synchroniser::is_today()` check because the `is_payment_upfront()` function checks if the product is synced to today [here](https://github.com/Automattic/woocommerce-subscriptions-core/blob/5.5.0/includes/class-wc-subscriptions-synchroniser.php#L548-L549).

Reply:
Thanks @james-allan